GUI/Types/Renderer/HitboxSetSceneNode.cs

```diff
@@ -0,0 +1,137 @@
+        {
+            this.animationController = animationController;
+
+            var boneIndexes = skeleton.Bones.Select((b, i) => (b, i))
+                                            .ToDictionary(p => p.b.Name.ToLowerInvariant(), p => p.i);
+
+            foreach (var pair in hitboxSets)
+            {
+                AddHitboxSet(pair.Key, pair.Value, boneIndexes);
+            }
+        }
+
+        private void AddHitboxSet(string name, Hitbox[] hitboxSet, Dictionary<string, int> boneIndexes)
+        {
+            var sceneNodes = new HitboxSceneNode[hitboxSet.Length];
+            var hitboxBoneIndexes = new int[hitboxSet.Length];
+
+            for (var i = 0; i < hitboxSet.Length; i++)
+            {
+                var hitbox = hitboxSet[i];
+                sceneNodes[i] = HitboxSceneNode.Create(Scene, hitbox);
+
+                if (string.IsNullOrEmpty(hitbox.BoneName) || !boneIndexes.TryGetValue(hitbox.BoneName.ToLowerInvariant(), out var boneIndex))
+                {
+                    hitboxBoneIndexes[i] = -1;
+                }
+                else
+                {
+                    hitboxBoneIndexes[i] = boneIndex;
+                }
+            }
+
+            var data = new HitboxSetData
+            {
+                HitboxSet = hitboxSet,
+                HitboxBoneIndexes = hitboxBoneIndexes,
+                SceneNodes = sceneNodes
+            };
+
+            hitboxSets.Add(name, data);
+        }
+
+        public void SetHitboxSet(string set)
+        {
+            if (set == null)
+            {
+                currentSet = null;
+                return;
+            }
+
+            currentSet = hitboxSets[set];
+        }
+
+        private static void UpdateHitboxSet(HitboxSetData hitboxSetData, Matrix4x4[] boneMatrices)
+        {
+            var hitboxSet = hitboxSetData.HitboxSet;
+            for (var i = 0; i < hitboxSet.Length; i++)
+            {
+                var shape = hitboxSetData.SceneNodes[i];
+                var hitbox = hitboxSet[i];
+                var boneId = hitboxSetData.HitboxBoneIndexes[i];
+                var targetTransform = boneId == -1 ? Matrix4x4.Identity : boneMatrices[boneId];
+
+                if (hitbox.TranslationOnly)
+                {
+                    Matrix4x4.Decompose(targetTransform, out _, out _, out var translation);
+                    shape.Transform = Matrix4x4.CreateTranslation(translation);
+                }
+                else
+                {
+                    shape.Transform = targetTransform;
+                }
+            }
+        }
+
+        public override void Update(Scene.UpdateContext context)
+        {
+            if (currentSet == null)
+            {
+                return;
+            }
+
+            LocalBoundingBox = new AABB(new Vector3(float.MinValue), new Vector3(float.MaxValue));
+
+            var boneMatrices = ArrayPool<Matrix4x4>.Shared.Rent(skeleton.Bones.Length);
```

Comment:
https://github.com/ValveResourceFormat/ValveResourceFormat/blob/36667375d95977993a312e35e9780ff8028b32f6/GUI/Types/Renderer/ModelSceneNode.cs#L98-L99

So the trick is to rent from <float> and marshall it to matrix4x4 so that it reuses the same buffer pool.